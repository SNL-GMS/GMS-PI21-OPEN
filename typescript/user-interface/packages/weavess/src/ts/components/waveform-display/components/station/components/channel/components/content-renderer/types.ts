import type { WeavessTypes } from '@gms/weavess-core';
import type React from 'react';

import type { PositionConverters } from '../../../../../../../../util/types';
import type { AnimationFrameOptions, updateMeasureWindow } from '../../../../../../types';

export interface ContentRendererProps {
  /** Station Id as string */
  stationId: string;

  /** The description */
  description?: string;

  /** The description label color */
  descriptionLabelColor?: string;

  /** Channel Id as string */
  channelId: string;

  /** Boolean is default channel */
  isDefaultChannel: boolean;

  /** waveform interval loaded and available to display */
  displayInterval: WeavessTypes.TimeRange;

  /** The ratio of the zoom interval divided by the total viewable interval. Unitless. */
  getZoomRatio: () => number;

  /** Web Workers */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  workerRpcs: any[];

  /** The signal detections */
  signalDetections: WeavessTypes.PickMarker[] | undefined;

  /** The predicted phases */
  predictedPhases: WeavessTypes.PickMarker[] | undefined;

  /** The theoretical phase windows */
  theoreticalPhaseWindows: WeavessTypes.TheoreticalPhaseWindow[] | undefined;

  /** Collection of markers */
  markers?: WeavessTypes.Markers;

  /** The selections */
  selections: WeavessTypes.Selections;

  /** (optional) callback events */
  events?: WeavessTypes.ChannelContentEvents;

  /** Configuration for weavess */
  initialConfiguration: WeavessTypes.Configuration;

  /** React elements to render in the content-renderer-content area */
  contentRenderers: React.ReactElement[];

  // Functions

  converters: PositionConverters;

  /** Ref to the html canvas element */
  canvasRef(): HTMLCanvasElement | null;

  /** Issues a rerender of the graphics */
  renderWaveforms(options?: AnimationFrameOptions): void;

  /**
   * Mouse move event
   *
   * @param e mouse event as React.MouseEvent<HTMLDivElement>
   */
  onMouseMove(e: React.MouseEvent<HTMLDivElement>): void;

  /**
   * Mouse down event
   *
   * @param e mouse event as React.MouseEvent<HTMLDivElement>
   */
  onMouseDown(e: React.MouseEvent<HTMLDivElement>): void;

  /**
   * onMouseUp event handler
   *
   * @param e mouse event as React.MouseEvent<HTMLDivElement>
   */
  onMouseUp(e: React.MouseEvent<HTMLDivElement>): void;

  /**
   * onContextMenu event handler
   *
   * @param e mouse event as React.MouseEvent<HTMLDivElement>
   */
  onContextMenu(e: React.MouseEvent<HTMLDivElement>): void;

  /**
   * onKeyDown event handler
   *
   * @param e mouse event as React.MouseEvent<HTMLDivElement>
   * @param timeForMouseXPosition the time for the current mouse position X
   */
  onKeyDown(e: React.KeyboardEvent<HTMLDivElement>): void;

  /**
   * (optional) Updates the measure window
   */
  updateMeasureWindow?: updateMeasureWindow;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface ContentRendererState {}
