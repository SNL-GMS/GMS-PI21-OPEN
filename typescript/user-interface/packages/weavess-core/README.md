# Weavess Util

Provides access to core WEAVESS features without a dependency on the window object. This allows these WEAVESS features to be used in web workers in external packages.

## Build Scripts
  * `clean:node_modules`: removes the node_modules directory
  * `clean:build`: removes all of the build directories, files, and artifacts
  * `clean`: cleans and removes all
  * `build:eslint`: runs eslint checks on the package for the source
  * `build:eslint:test`: runs eslint checks on the package for tests
  * `build`: runs the build of the package for the source
  * `build:test`: runs the build of the package for tests
  * `watch`: runs the build and watches for changes to recompile
  * `docs`: generates the package source documentation
  * `sonar`: runs sonar lint checks
  * `test`: builds and runs the package jest tests as development
  * `test:dev:jest`: runs the package jest tests as development
  * `test:prod:jest`: runs the package jest tests as production
  * `test:dev`: builds and runs the package jest tests as development
  * `test:prod`: builds and runs the package jest tests as production 
  * `version`: returns the version of the package

## Documentation

To generate HTML documentation files:

```bash
yarn docs
```

## Building

```bash
[.../weavess-core] $ yarn build
```

## Generating Binaries

```bash
[.../weavess-core] $ yarn build
```
