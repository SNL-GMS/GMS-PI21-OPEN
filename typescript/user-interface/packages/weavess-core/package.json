{
  "name": "@gms/weavess-core",
  "version": "0.1.0",
  "description": "WEAVESS Core",
  "license": "UNLICENSED",
  "author": "GMS Development Team",
  "sideEffects": false,
  "module": "./lib/weavess-core.js",
  "typings": "./lib/weavess-core.d.ts",
  "files": [
    "lib"
  ],
  "scripts": {
    "clean:node_modules": "yarn rimraf node_modules/",
    "clean:build": "yarn rimraf lib/ dist/ docs/ coverage/ bundle-analyzer/ artifacts/ .jest-cache/ .tsbuildinfo .tsbuildinfo-test .eslintcache .eslintcache-test junit.xml",
    "clean": "npm-run-all -s clean:build clean:node_modules",
    "build:eslint": "yarn eslint -c .eslintrc-config.yaml 'src/**/*.{ts,tsx}' --quiet --cache --cache-file .eslintcache",
    "build:eslint:test": "yarn eslint -c .eslintrc-config-test.yaml '{__tests__,__mocks__,__integration-tests__}/**/*.{ts,tsx}' --quiet --cache --cache-file .eslintcache-test",
    "build": "yarn tsc -b ./tsconfig-build.json",
    "build:test": "yarn tsc -b ./tsconfig-test.json",
    "watch": "yarn tsc -b ./tsconfig-build.json -watch",
    "docs": "yarn clean:docs && typedoc --out docs ./src/ --name weavess-core --mode file --readme README.md --target ES6 --includeDeclarations --excludeExternals --excludeNotExported --hideGenerator",
    "sonar": "echo \"run 'yarn sonar' from the top-level directory\"",
    "test": "yarn test:dev",
    "test:dev:jest": "jest --config jest.config.js --maxWorkers=50% --verbose --bail",
    "test:prod:jest": "jest --config jest.config.js --maxWorkers=4 --silent --bail --coverageReporters=lcovonly --reporters=default --reporters=jest-junit",
    "test:dev": "npm-run-all -s build:test test:dev:jest",
    "test:prod": "npm-run-all -s build:test test:prod:jest",
    "version": "echo $npm_package_version"
  },
  "dependencies": {
    "csv-parse": "4.8.2",
    "d3": "5.5.0",
    "immutable": "4.0.0-rc.12",
    "lodash": "4.17.21",
    "math-random": "2.0.1",
    "moment": "2.26.0",
    "react": "18.1.0",
    "react-dom": "18.1.0",
    "uuid": "3.3.3"
  },
  "devDependencies": {
    "@blueprintjs/eslint-plugin": "2.2.0",
    "@testing-library/jest-dom": "5.16.4",
    "@testing-library/react": "13.2.0",
    "@types/d3": "4.12.0",
    "@types/faker": "4.1.8",
    "@types/jest": "27.0.1",
    "@types/lodash": "4.14.149",
    "@types/react": "18.0.8",
    "@types/react-dom": "18.0.3",
    "@types/react-router-dom": "5.3.3",
    "@types/uuid": "3.4.6",
    "@typescript-eslint/eslint-plugin": "5.33.0",
    "@typescript-eslint/parser": "5.33.0",
    "copyfiles": "2.1.1",
    "eslint": "8.22.0",
    "eslint-config-airbnb": "19.0.4",
    "eslint-config-prettier": "8.5.0",
    "eslint-import-resolver-typescript": "3.4.1",
    "eslint-plugin-import": "2.26.0",
    "eslint-plugin-jest": "26.8.3",
    "eslint-plugin-jsdoc": "39.3.6",
    "eslint-plugin-jsx-a11y": "6.6.1",
    "eslint-plugin-no-null": "1.0.2",
    "eslint-plugin-prefer-arrow": "1.2.3",
    "eslint-plugin-prettier": "4.2.1",
    "eslint-plugin-react": "7.30.1",
    "eslint-plugin-react-hooks": "4.6.0",
    "eslint-plugin-simple-import-sort": "7.0.0",
    "eslint-plugin-unicorn": "43.0.2",
    "eslint-plugin-workspaces": "0.7.0",
    "faker": "4.1.0",
    "jest": "28.1.0",
    "jest-canvas-mock": "2.4.0",
    "jest-cli": "28.1.0",
    "jest-environment-jsdom": "28.1.0",
    "jest-fetch-mock": "3.0.3",
    "jsdom": "19.0.0",
    "jsdom-global": "3.0.2",
    "jsdom-worker": "0.1.0",
    "node-fetch": "2.6.0",
    "npm-run-all": "4.1.5",
    "prettier": "2.1.2",
    "rimraf": "3.0.0",
    "ts-jest": "28.0.1",
    "ts-node": "9.1.1",
    "typedoc": "0.15.3",
    "typescript": "4.6.4"
  },
  "jest-junit": {
    "suiteNameTemplate": "weavess-core/{filename}",
    "classNameTemplate": "weavess-core/{filename}/{classname}",
    "titleTemplate": "{title}"
  }
}
