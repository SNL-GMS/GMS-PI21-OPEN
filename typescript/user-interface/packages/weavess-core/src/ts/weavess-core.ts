import * as WeavessConstants from './constants';
import { Messages } from './messages';
import * as WeavessTypes from './types';
import * as WeavessUtil from './util';

export { WeavessConstants, Messages as WeavessMessages, WeavessTypes, WeavessUtil };
