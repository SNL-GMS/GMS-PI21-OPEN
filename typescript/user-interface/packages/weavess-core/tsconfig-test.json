{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "tsBuildInfoFile": "./.tsbuildinfo-test"
  },
  "references": []
}
