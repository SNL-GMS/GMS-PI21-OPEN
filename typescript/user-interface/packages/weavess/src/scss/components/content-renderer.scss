.contentrenderer {
  display: flex;
  height: inherit;
  flex: 1 1 auto;
  &:focus {
    outline: none;
  }
  border-right: 1px solid;

  .contentrenderer-content {
    position: absolute;
    height: inherit;
    width: calc(100% - 184px);
    overflow: hidden;

    &--sticky {
      position: sticky;
      width: calc(100vw - 184px);
      left: 184px;
      right: 0;
      overflow: unset;
    }

    &-description-label {
      display: flex;
      text-align: right;
      height: 100%;
      width: 100%;
      bottom: 0px;
      margin-left: auto;
      font-size: 0.7rem;
      white-space: nowrap;
    }
    &__description {
      padding-right: 6px;
      margin-left: auto;
      margin-top: auto;
    }

    &-drag-indicator {
      position: absolute;
      top: 0px;
      bottom: 0px;
      border-left: 1px solid red;
      display: none;
    }

    &-markers {
      pointer-events: none;
      position: absolute;
      width: 100%;
      height: 100%;
      top: 0;
    }
  }
}
