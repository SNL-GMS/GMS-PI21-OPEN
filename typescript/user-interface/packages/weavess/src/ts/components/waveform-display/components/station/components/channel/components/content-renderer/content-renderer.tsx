/* eslint-disable react/default-props-match-prop-types */
/* eslint-disable react/destructuring-assignment */
import { WeavessConstants, WeavessTypes } from '@gms/weavess-core';
import React from 'react';

import {
  memoizedCreateMoveableMarkers,
  memoizedCreateSelectionWindowMarkers,
  memoizedCreateVerticalMarkers
} from '../../../../../markers';
import { PredictedPhases, SignalDetections, TheoreticalPhases } from './components';
import type { ContentRendererProps, ContentRendererState } from './types';

/**
 * Content renderer component responsible for rendering the main content of a channel.
 */
export class ContentRenderer extends React.PureComponent<
  React.PropsWithChildren<ContentRendererProps>,
  ContentRendererState
> {
  /** Default channel props, if not provided */
  // eslint-disable-next-line react/static-property-placement
  public static readonly defaultProps: WeavessTypes.ChannelDefaultConfiguration = {
    displayType: [WeavessTypes.DisplayType.LINE],
    pointSize: 2,
    color: '#4580E6'
  };

  /** Ref to the element where this channel will be rendered */
  public containerRef: HTMLElement | null;

  /** Ref to the element where this description label will be rendered */
  // eslint-disable-next-line
  public descriptionLabelRef: HTMLElement | null;

  /** Ref to drag indicator element */
  private dragIndicatorRef: HTMLDivElement | null;

  private readonly resizeObserver: ResizeObserver;

  /**
   * Constructor
   *
   * @param props props as ContentRendererProps
   */
  public constructor(props: ContentRendererProps) {
    super(props);
    this.resizeObserver = new ResizeObserver(() => this.forceUpdate());
    this.state = {};
  }

  public componentDidMount(): void {
    const canvasRef = this.props.canvasRef();
    if (canvasRef) {
      this.resizeObserver.observe(canvasRef);
    }
  }

  public componentWillUnmount(): void {
    const canvasRef = this.props.canvasRef();
    if (this.resizeObserver && canvasRef) {
      this.resizeObserver.unobserve(canvasRef);
    }
  }

  // eslint-disable-next-line react/sort-comp
  public render(): JSX.Element {
    const isSelected =
      this.props.selections.channels &&
      this.props.selections.channels.indexOf(this.props.channelId) > -1;
    const labelWidthPx =
      this.props.initialConfiguration?.labelWidthPx ?? WeavessConstants.DEFAULT_LABEL_WIDTH_PIXELS;
    const anySignalDetectionsSelected = this.props.signalDetections
      ? this.props.signalDetections.reduce((anySelected, sd) => anySelected || sd.isSelected, false)
      : false;
    return (
      // eslint-disable-next-line jsx-a11y/mouse-events-have-key-events, jsx-a11y/no-static-element-interactions
      <div
        className={`contentrenderer${
          anySignalDetectionsSelected ? ' contentrenderer--selected' : ''
        }`}
        style={{
          backgroundColor: isSelected ? 'rgba(150,150,150,0.2)' : 'initial'
        }}
        ref={ref => {
          if (ref) {
            this.containerRef = ref;
          }
        }}
        // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
        tabIndex={0}
        onKeyDown={this.props.onKeyDown}
        onContextMenu={this.props.onContextMenu}
        onMouseMove={this.props.onMouseMove}
        onMouseDown={this.props.onMouseDown}
        onMouseUp={this.props.onMouseUp}
      >
        {this.props.children}
        <div
          className="contentrenderer-content contentrenderer-content--sticky"
          style={{
            width: this.props.canvasRef()?.clientWidth ?? `calc(100% - ${labelWidthPx})`
          }}
        >
          {this.props.description ? (
            <div
              ref={ref => {
                this.descriptionLabelRef = ref;
              }}
              className="contentrenderer-content-description-label"
              data-cy="filtered-channel-label"
              data-cy-color={this.props.descriptionLabelColor}
              style={{
                color: this.props.descriptionLabelColor
              }}
            >
              <span className="contentrenderer-content__description">{this.props.description}</span>
            </div>
          ) : undefined}
        </div>
        <div
          className="contentrenderer-content"
          style={{
            left: labelWidthPx,
            width: `calc(100% - ${labelWidthPx}px)`
          }}
          data-cy-station-waveform={this.props.channelId}
        >
          <div
            ref={ref => {
              this.dragIndicatorRef = ref;
            }}
            className="contentrenderer-content-drag-indicator"
          />
          {...this.props.contentRenderers}
          <div className="contentrenderer-content-markers">{this.createAllMarkers(this.props)}</div>
          <SignalDetections
            stationId={this.props.stationId}
            channelId={this.props.channelId}
            signalDetections={this.props.signalDetections}
            isDefaultChannel={this.props.isDefaultChannel}
            displayInterval={this.props.displayInterval}
            events={this.props.events}
            getTimeSecsForClientX={this.getTimeSecsForClientX}
            toggleDragIndicator={this.toggleDragIndicator}
            positionDragIndicator={this.positionDragIndicator}
          />
          <PredictedPhases
            stationId={this.props.stationId}
            channelId={this.props.channelId}
            predictedPhases={this.props.predictedPhases}
            isDefaultChannel={this.props.isDefaultChannel}
            displayInterval={this.props.displayInterval}
            selectedPredictedPhases={this.props.selections.predictedPhases}
            events={this.props.events}
            getTimeSecsForClientX={this.getTimeSecsForClientX}
            toggleDragIndicator={this.toggleDragIndicator}
            positionDragIndicator={this.positionDragIndicator}
          />
          <TheoreticalPhases
            stationId={this.props.stationId}
            theoreticalPhaseWindows={this.props.theoreticalPhaseWindows}
            isDefaultChannel={this.props.isDefaultChannel}
            displayInterval={this.props.displayInterval}
            events={this.props.events}
            getTimeSecsForClientX={this.getTimeSecsForClientX}
            toggleDragIndicator={this.toggleDragIndicator}
            positionDragIndicator={this.positionDragIndicator}
          />
        </div>
      </div>
    );
  }

  /**
   * Creates all of the markers.
   *
   * @param props the content renderer props
   *
   * @returns an array JSX elements
   */
  private readonly createAllMarkers = (props: ContentRendererProps): JSX.Element[] => [
    ...memoizedCreateVerticalMarkers(
      props.displayInterval.startTimeSecs,
      props.displayInterval.endTimeSecs,
      props.markers ? props.markers.verticalMarkers : undefined
    ),
    ...memoizedCreateMoveableMarkers(
      props.displayInterval.startTimeSecs,
      props.displayInterval.endTimeSecs,
      props.markers ? props.markers.moveableMarkers : undefined,
      props.getZoomRatio,
      () => (this.containerRef ? this.containerRef.clientWidth : 0),
      () => (this.containerRef ? this.containerRef.clientWidth : 0),
      props.events
        ? (marker: WeavessTypes.Marker) => {
            if (props.events && props.events.onUpdateMarker) {
              props.events.onUpdateMarker(props.channelId, marker);
            }
          }
        : undefined,
      0
    ),
    ...memoizedCreateSelectionWindowMarkers(
      props.displayInterval.startTimeSecs,
      props.displayInterval.endTimeSecs,
      props.markers ? props.markers.selectionWindows : undefined,
      props.getZoomRatio,
      () => props.canvasRef(),
      () => (this.containerRef ? this.containerRef.clientWidth : 0),
      () => (this.containerRef ? this.containerRef.clientWidth : 0),
      // eslint-disable-next-line @typescript-eslint/unbound-method
      props.converters.computeTimeSecsForMouseXFractionalPosition,
      // eslint-disable-next-line @typescript-eslint/unbound-method
      props.onMouseMove,
      // eslint-disable-next-line @typescript-eslint/unbound-method
      props.onMouseDown,
      // eslint-disable-next-line @typescript-eslint/unbound-method
      props.onMouseUp,
      props.events
        ? (selection: WeavessTypes.SelectionWindow) => {
            if (props.events && props.events.onMoveSelectionWindow) {
              props.events.onMoveSelectionWindow(props.channelId, selection);
            }
          }
        : undefined,
      props.events
        ? (selection: WeavessTypes.SelectionWindow) => {
            if (props.events && props.events.onUpdateSelectionWindow) {
              props.events.onUpdateSelectionWindow(props.channelId, selection);
            }
          }
        : undefined,
      props.events
        ? (selection: WeavessTypes.SelectionWindow, timeSecs: number) => {
            if (props.events && props.events.onClickSelectionWindow) {
              props.events.onClickSelectionWindow(props.channelId, selection, timeSecs);
            }
          }
        : undefined,
      0
    )
  ];

  /**
   * Returns the time in seconds for the given clientX.
   *
   * @param clientX The clientX
   *
   * @returns The time in seconds; undefined if clientX is
   * out of the channel's bounds on screen.
   */
  private readonly getTimeSecsForClientX = (clientX: number): number | undefined => {
    const canvasRef = this.props.canvasRef();

    if (!this.containerRef || !canvasRef) return undefined;

    const offset = canvasRef.getBoundingClientRect();
    // eslint-disable-next-line consistent-return
    if (clientX < offset.left && clientX > offset.right) return undefined;

    // position in [0,1] in the current channel bounds.
    const position = (clientX - offset.left) / offset.width;
    // eslint-disable-next-line consistent-return
    return this.props.converters.computeTimeSecsForMouseXFractionalPosition(position);
  };

  /**
   * Toggle display of the drag indicator for this channel
   *
   * @param show True to show drag indicator
   * @param color The color of the drag indicator
   */
  private readonly toggleDragIndicator = (show: boolean, color: string): void => {
    if (!this.dragIndicatorRef) return;

    this.dragIndicatorRef.style.borderColor = color;
    this.dragIndicatorRef.style.display = show ? 'initial' : 'none';
  };

  /**
   * Set the position for the drag indicator
   *
   * @param clientX The clientX
   */
  private readonly positionDragIndicator = (clientX: number): void => {
    if (!this.containerRef || !this.dragIndicatorRef) return;

    const fracToPct = 100;
    const boundingRect = this.containerRef.getBoundingClientRect();
    // position in [0,1] in the current channel bounds.
    const position = (clientX - boundingRect.left) / boundingRect.width;
    this.dragIndicatorRef.style.left = `${position * fracToPct}%`;
  };
}
