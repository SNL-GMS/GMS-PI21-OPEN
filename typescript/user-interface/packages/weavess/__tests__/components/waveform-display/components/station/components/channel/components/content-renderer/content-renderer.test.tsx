/* eslint-disable @typescript-eslint/no-magic-numbers */
/* eslint-disable react/jsx-props-no-spreading */
import { WeavessTypes } from '@gms/weavess-core';
import { render } from '@testing-library/react';
import * as Enzyme from 'enzyme';
import * as React from 'react';

import { ContentRenderer } from '../../../../../../../../../src/ts/components/waveform-display/components/station/components/channel/components/content-renderer/content-renderer';
import type { ContentRendererProps } from '../../../../../../../../../src/ts/components/waveform-display/components/station/components/channel/components/content-renderer/types';

const initialConfiguration: WeavessTypes.Configuration = {
  shouldRenderWaveforms: true,
  shouldRenderSpectrograms: false,
  hotKeys: {
    amplitudeScale: 'KeyS',
    amplitudeScaleSingleReset: 'Alt+KeyY',
    amplitudeScaleReset: 'Alt+Shift+KeyY',
    maskCreate: 'KeyM'
  },
  waveformDimPercent: 0.75,
  backgroundColor: '#182026',
  defaultChannel: {
    disableMeasureWindow: true,
    disableMaskModification: true
  },
  nonDefaultChannel: {
    disableMeasureWindow: true,
    disableMaskModification: false
  }
};

const props: ContentRendererProps = {
  initialConfiguration,
  channelId: 'channel-id',
  stationId: 'station-id',
  displayInterval: {
    startTimeSecs: 300,
    endTimeSecs: 700
  },
  getZoomRatio: jest.fn().mockReturnValue(0.5),
  isDefaultChannel: true,
  predictedPhases: [],
  selections: {
    channels: [],
    predictedPhases: [],
    signalDetections: []
  },
  signalDetections: [],
  theoreticalPhaseWindows: [],
  workerRpcs: [],
  contentRenderers: [],
  description: 'description',
  descriptionLabelColor: '#ff000',
  events: {
    onChannelClick: jest.fn(),
    onClickSelectionWindow: jest.fn(),
    onContextMenu: jest.fn(),
    onMaskClick: jest.fn(),
    onMaskContextClick: jest.fn(),
    onMaskCreateDragEnd: jest.fn(),
    onMeasureWindowUpdated: jest.fn(),
    onMoveSelectionWindow: jest.fn(),
    onPredictivePhaseClick: jest.fn(),
    onPredictivePhaseContextMenu: jest.fn(),
    onPredictivePhaseDragEnd: jest.fn(),
    onSignalDetectionClick: jest.fn(),
    onSignalDetectionContextMenu: jest.fn(),
    onSignalDetectionDragEnd: jest.fn(),
    onUpdateMarker: jest.fn(),
    onUpdateSelectionWindow: jest.fn()
  },
  markers: {
    moveableMarkers: [],
    selectionWindows: [],
    verticalMarkers: []
  },
  canvasRef: jest.fn(),
  converters: {
    computeFractionOfCanvasFromMouseXPx: jest.fn(() => 88),
    computeTimeSecsForMouseXFractionalPosition: jest.fn(() => 88),
    computeTimeSecsFromMouseXPixels: jest.fn(() => 88)
  },
  onContextMenu: jest.fn(),
  onKeyDown: jest.fn(),
  onMouseDown: jest.fn(),
  onMouseMove: jest.fn(),
  onMouseUp: jest.fn(),
  renderWaveforms: jest.fn(),
  updateMeasureWindow: jest.fn()
};

const wrapper = Enzyme.mount(<ContentRenderer {...props} />);
const instance: any = wrapper.find(ContentRenderer).instance();

describe('Weavess Content Renderer', () => {
  it('to be defined', () => {
    expect(ContentRenderer).toBeDefined();
  });

  it('renders', () => {
    const { container } = render(<ContentRenderer {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('create all markers', () => {
    const p: ContentRendererProps = {
      ...props,
      markers: {
        moveableMarkers: [
          {
            id: 'moveable',
            timeSecs: 40,
            color: 'ff000',
            lineStyle: WeavessTypes.LineStyle.DASHED,
            minTimeSecsConstraint: 4,
            maxTimeSecsConstraint: 200
          }
        ],
        selectionWindows: [
          {
            id: 'selection',
            color: '00ff00',
            isMoveable: true,
            startMarker: {
              id: 'start',
              color: '#ff0000',
              lineStyle: WeavessTypes.LineStyle.SOLID,
              timeSecs: 10,
              minTimeSecsConstraint: 1,
              maxTimeSecsConstraint: 200
            },
            endMarker: {
              id: 'end',
              color: '#ff0000',
              lineStyle: WeavessTypes.LineStyle.SOLID,
              timeSecs: 80,
              minTimeSecsConstraint: 1,
              maxTimeSecsConstraint: 200
            }
          }
        ],
        verticalMarkers: [
          {
            id: 'vertical',
            color: '0000ff',
            lineStyle: WeavessTypes.LineStyle.DASHED,
            timeSecs: 120,
            minTimeSecsConstraint: 0,
            maxTimeSecsConstraint: 300
          }
        ]
      }
    };
    const { container } = render(<ContentRenderer {...p} />);
    expect(container).toMatchSnapshot();
  });

  it('getTimeSecsForClientX', () => {
    let value = instance.getTimeSecsForClientX(43);
    expect(value).toEqual(undefined);

    const computeTimeSecsForMouseXFractionalPosition = jest.fn(() => 88);
    const content = Enzyme.mount(
      <ContentRenderer
        {...props}
        converters={{
          computeFractionOfCanvasFromMouseXPx: jest.fn(() => 88),
          computeTimeSecsForMouseXFractionalPosition,
          computeTimeSecsFromMouseXPixels: jest.fn(() => 88)
        }}
        canvasRef={() =>
          ({
            getBoundingClientRect: () => ({
              right: 9,
              left: 11,
              width: 100
            })
          } as any)
        }
      />
    );
    const contentInstance: any = content.find(ContentRenderer).instance();
    value = contentInstance.getTimeSecsForClientX(10);
    expect(value).toEqual(undefined);

    value = contentInstance.getTimeSecsForClientX(18);
    expect(computeTimeSecsForMouseXFractionalPosition).toHaveBeenCalledTimes(1);
    expect(value).toEqual(88);
  });

  it('toggleDragIndicator', () => {
    instance.toggleDragIndicator(true, '000000');
    expect(instance.dragIndicatorRef.style.display).toEqual('initial');
    instance.toggleDragIndicator(false, '000000');
    expect(instance.dragIndicatorRef.style.display).toEqual('none');

    const content = Enzyme.mount(<ContentRenderer {...props} />);
    const contentInstance: any = content.find(ContentRenderer).instance();
    contentInstance.dragIndicatorRef = undefined;
    contentInstance.toggleDragIndicator(true, '000000');
  });

  it('positionDragIndicator', () => {
    const content = Enzyme.mount(<ContentRenderer {...props} />);

    const getBoundingClientRect = jest.fn(() => ({
      left: 44,
      right: 12,
      width: 100
    }));
    const contentInstance: any = content.find(ContentRenderer).instance();
    expect(contentInstance.dragIndicatorRef.style.display).toEqual('');

    contentInstance.containerRef = {
      getBoundingClientRect
    };
    contentInstance.positionDragIndicator(48);
    expect(contentInstance.dragIndicatorRef.style.left).toEqual('4%');
    expect(getBoundingClientRect).toHaveBeenCalledTimes(1);

    contentInstance.dragIndicatorRef = undefined;
    contentInstance.positionDragIndicator(48);
    expect(getBoundingClientRect).toHaveBeenCalledTimes(1);
  });
});
