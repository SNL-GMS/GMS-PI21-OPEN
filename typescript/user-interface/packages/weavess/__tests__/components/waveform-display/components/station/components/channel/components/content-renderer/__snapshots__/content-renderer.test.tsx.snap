// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weavess Content Renderer create all markers 1`] = `
<div>
  <div
    class="contentrenderer"
    tabindex="0"
  >
    <div
      class="contentrenderer-content contentrenderer-content--sticky"
      style="width: calc(100% - 184);"
    >
      <div
        class="contentrenderer-content-description-label"
        data-cy="filtered-channel-label"
        data-cy-color="#ff000"
      >
        <span
          class="contentrenderer-content__description"
        >
          description
        </span>
      </div>
    </div>
    <div
      class="contentrenderer-content"
      data-cy-station-waveform="channel-id"
      style="left: 184px; width: calc(100% - 184px);"
    >
      <div
        class="contentrenderer-content-drag-indicator"
      />
      <div
        class="contentrenderer-content-markers"
      />
    </div>
  </div>
</div>
`;

exports[`Weavess Content Renderer renders 1`] = `
<div>
  <div
    class="contentrenderer"
    tabindex="0"
  >
    <div
      class="contentrenderer-content contentrenderer-content--sticky"
      style="width: calc(100% - 184);"
    >
      <div
        class="contentrenderer-content-description-label"
        data-cy="filtered-channel-label"
        data-cy-color="#ff000"
      >
        <span
          class="contentrenderer-content__description"
        >
          description
        </span>
      </div>
    </div>
    <div
      class="contentrenderer-content"
      data-cy-station-waveform="channel-id"
      style="left: 184px; width: calc(100% - 184px);"
    >
      <div
        class="contentrenderer-content-drag-indicator"
      />
      <div
        class="contentrenderer-content-markers"
      />
    </div>
  </div>
</div>
`;
